app/models/category_tag_stat.rb

```diff
@@ -0,0 +1,61 @@
+class CategoryTagStat < ActiveRecord::Base
+  belongs_to :category
+  belongs_to :tag
+
+  def self.topic_moved(topic, from_category_id, to_category_id)
+    if from_category_id
+      self.where(tag_id: topic.tags.map(&:id), category_id: from_category_id)
+        .where('topic_count > 0')
+        .update_all('topic_count = topic_count - 1')
+    end
+
+    if to_category_id
+      sql = <<~SQL
+        UPDATE #{self.table_name}
+           SET topic_count = topic_count + 1
+         WHERE tag_id in (:tag_ids)
+           AND category_id = :category_id
+     RETURNING tag_id
+      SQL
+
+      tag_ids = topic.tags.map(&:id)
+      updated_tag_ids = self.exec_sql(sql, tag_ids: tag_ids, category_id: to_category_id).map { |row| row['tag_id'] }
+
+      (tag_ids - updated_tag_ids).each do |tag_id|
+        CategoryTagStat.create(tag_id: tag_id, category_id: to_category_id, topic_count: 1)
```

Comment:
I think it'll be better to use `create!` if we're not checking for the return value of `create`